File: bots/server_module/game_rules/game_rules.py
from server_module.game_rules.board_cards import BoardCards
from server_module.game_rules.board_tile import BoardTile
from server_module.game_rules.military_unit_info import MilitaryUnitInfo
from server_module.game_state.house_card import HouseCard
from server_module.game_state.military_unit_type import MilitaryUnitType
from server_module.game_state.order import Order
from server_module.game_state.order_type import OrderType


class GameRules:
    def __init__(
            self,
            board: list[BoardTile],
            kings_court_stars: list[int],
            supply_usage: list[list[int]],
            max_armies: dict[MilitaryUnitType, int],
            board_cards: BoardCards,
            loaded_orders: dict[OrderType, list[Order]],
            house_cards: list[HouseCard],
            military_units: dict[MilitaryUnitType, MilitaryUnitInfo]
    ):
        self.board = board
        self.kings_court_stars = kings_court_stars
        self.supply_usage = supply_usage
        self.max_armies = max_armies
        self.board_cards = board_cards
        self.loaded_orders = loaded_orders
        self.house_cards = house_cards
        self.military_units = military_units

    @classmethod
    def from_json(cls, json):
        return cls(
            list(BoardTile.from_json(bt) for bt in json['board']),
            json['kingsCourtStars'],
            json['supplyUsage'],
            {MilitaryUnitType.from_str(mut): num for mut, num in json['maxArmies'].items()},
            BoardCards.from_json(json['boardCards']),
            {OrderType.from_str(ot): list(Order.from_json(o) for o in od) for ot, od in json['loadedOrders'].items()},
            list(HouseCard.from_json(hc) for hc in json['houseCards']),
            {MilitaryUnitType.from_str(mut): MilitaryUnitInfo.from_json(mui) for mut, mui in json['militaryUnits'].items()}
        )

File: game-logic-core/scala3MainProject/src/main/scala/fwc/game/GameRules.scala
package fwc.game

import fwc.JsonSerializable
import fwc.game.board.{Board, MilitaryUnitType}
import fwc.game.eventsPhase.cards.BoardCards
import fwc.game.planningPhase.{Order, OrderType}
import fwc.gameLoading.HouseCard
import ujson.Value

case class GameRules(
                    board: Board,
                    kingsCourtStars: Vector[Int],
                    supplyUsage: Vector[Seq[Int]],
                    maxArmies: Map[MilitaryUnitType, Int],
                    boardCards: BoardCards,
                    boardCardsForClient: BoardCards,
                    loadedOrders: Map[OrderType, Seq[Order]],
                    houseCards: Seq[HouseCard]
                    ) 
  extends JsonSerializable {
  override def toJson: Value = ujson.Obj(
    "board" -> board.map(_.toJson),
    "kingsCourtStars" -> kingsCourtStars,
    "supplyUsage" -> supplyUsage,
    "maxArmies" -> maxArmies.map((mut, n) => mut.toString -> n),
    "boardCards" -> boardCardsForClient.toRulesJson,
    "loadedOrders" -> loadedOrders.map((ot, ods) => ot.toString -> ods.map(_.toJson)),
    "houseCards" -> houseCards.map(_.toJson),
    "militaryUnits" -> ujson.Obj(
      MilitaryUnitType.Footmen.toString -> MilitaryUnitType.Footmen.toJson,
      MilitaryUnitType.Knights.toString -> MilitaryUnitType.Knights.toJson,
      MilitaryUnitType.Ships.toString -> MilitaryUnitType.Ships.toJson,
      MilitaryUnitType.SiegeEngines.toString -> MilitaryUnitType.SiegeEngines.toJson
    )
  )
}
